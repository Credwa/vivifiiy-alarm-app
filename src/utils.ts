import { meridiem, twelveHrTime } from '@/types';
import { smallScreenWidthBreakpoint, largeScreenWidthBreakpoint } from '@/constants/Values';
import { Dimensions } from 'react-native';
/**
 * ConvertTo24Hour
 *
 * Convert 12 hour time to military time (24 hour)
 *
 * @param time
 *
 */
export const convertTo24Hour = (time: twelveHrTime) => {
  const newTime: twelveHrTime = { hour: time.hour, minute: time.minute, meridiem: time.meridiem };
  if (time.meridiem === 'AM' && time.hour === '12') {
    newTime.hour = '00';
  } else if (time.meridiem === 'PM' && time.hour !== '12') {
    newTime.hour = (Number(time.hour) + 12).toString();
  }
  return newTime;
};

/**
 * generateTimerArray
 *
 * Generates an array of continuous numbers starting from 0 (if included) to max. Option to repeat (append generate array to itself) a specific number of time
 *
 * @param max
 * @param zeroIncluded
 * @param startingRepeats
 *
 */
export const generateTimerArray = (max: number, zeroIncluded: boolean, startingRepeats?: number) => {
  let tempArr: string[] = [];
  for (let index = zeroIncluded ? 0 : 1; index <= max; index++) {
    tempArr.push(index.toLocaleString('en-IN', { minimumIntegerDigits: 2 }));
  }
  if (startingRepeats) {
    let originalArr = tempArr;
    for (let index = 0; index < startingRepeats; index++) {
      tempArr = [...tempArr, ...originalArr];
    }
  }
  return tempArr;
};

/**
 * findSelectedAlarmViewIndex
 *
 * returns the index for matching initial alarm view value
 *
 * @param data
 * @param initValue
 * @param startingIndex
 */
export const findSelectedAlarmViewIndex = (
  data: meridiem[] | string[],
  initValue: meridiem | string,
  startingIndex: number
) => {
  let index = Math.floor(data.length / 2);
  for (let i = startingIndex; i < data.length; i++) {
    if (data[i] === initValue) {
      index = i;
    }
  }
  return index;
};

/**
 * debounce
 *
 * @param func
 * @param waitFor
 */
export const debounce = <F extends (...args: any[]) => any>(func: F, waitFor: number) => {
  let timeout: ReturnType<typeof setTimeout> | null = null;

  const debounced = (...args: Parameters<F>) => {
    if (timeout !== null) {
      clearTimeout(timeout);
      timeout = null;
    }
    timeout = setTimeout(() => func(...args), waitFor);
  };

  return debounced as (...args: Parameters<F>) => ReturnType<F>;
};

/**
 * resize
 *
 * Resizes items depending on screen size
 *
 * @param standard
 * @param small
 * @param large
 */
export function resize<T>(standard: T, small: T, large: T) {
  const windowWidth = Dimensions.get('window').width;
  return windowWidth < smallScreenWidthBreakpoint
    ? small
    : windowWidth > largeScreenWidthBreakpoint
    ? large
    : standard;
}

/**
 * truncateString
 *
 * Returns a truncated string with ... concated
 *
 * @param str
 * @param num
 */
export function truncateString(str: string, maxLen: number) {
  if (str.length <= maxLen) {
    return str;
  }

  return str.slice(0, maxLen) + '...';
}
