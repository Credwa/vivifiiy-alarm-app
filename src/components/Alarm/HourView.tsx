import React, { useEffect, useState } from 'react';
import { NativeSyntheticEvent, NativeScrollEvent } from 'react-native';
import ScrollSelector from '@/components/ScrollSelector';
import Colors from '@/constants/Colors';
import useStore from '@/store';
import { findSelectedAlarmViewIndex, resize } from '@/utils';
import { FontName } from '../VivText';
interface SnapScrollProps {
  data: string[];
  initValue: string;
}

export default function HourView({ data, initValue }: SnapScrollProps) {
  const [listState, setEndOfList] = useState(data);
  const [timesExtended, setTimesExtended] = useState(0);
  const updateHourOnChange = useStore((state) => state.setHour);
  const onEndReached = (event: NativeSyntheticEvent<NativeScrollEvent> | NativeScrollEvent) => {
    if (timesExtended <= 2) {
      setEndOfList([...listState, ...data]);
      setTimesExtended(timesExtended + 1);
    }
  };
  useEffect(() => {
    updateHourOnChange(initValue);
  }, []);

  return (
    <ScrollSelector
      dataSource={listState}
      overrideFontName={resize<FontName[]>(['Title1', 'Title2'], ['Title3', 'Title4'], ['TitleBig1', 'TitleBig2'])}
      selectedIndex={findSelectedAlarmViewIndex(data, initValue, 11)}
      onValueChange={(data) => {
        updateHourOnChange(data);
      }}
      wrapperHeight={resize<number>(200, 175, 300)}
      wrapperWidth={resize<number>(60, 45, 150)}
      onEndReached={onEndReached}
      itemHeight={resize<number>(75, 65, 110)}
      highlightColor={Colors.greyLight3}
      highlightBorderWidth={1}
    />
  );
}
