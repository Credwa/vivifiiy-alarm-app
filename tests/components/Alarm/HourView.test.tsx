import React from 'react';
import renderer from 'react-test-renderer';
import HourView from '@/components/Alarm/HourView';
import { generateTimerArray } from '@/utils';
jest.useFakeTimers();
describe('<HourView />', () => {
  it(`renders correctly`, () => {
    const tree = renderer.create(<HourView data={generateTimerArray(12, false, 2)} initValue={'08'} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
