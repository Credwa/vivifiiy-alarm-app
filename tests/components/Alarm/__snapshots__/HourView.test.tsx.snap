// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HourView /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "backgroundColor": null,
      "height": 200,
      "overflow": "hidden",
      "width": 60,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#B1C2D1",
        "borderBottomWidth": 1,
        "borderTopColor": "#B1C2D1",
        "borderTopWidth": 1,
        "height": 75,
        "position": "absolute",
        "top": 62.5,
        "width": "100%",
      }
    }
  />
  <RCTScrollView
    bounces={false}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    scrollEventThrottle={400}
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "flex": 1,
            "height": 62.5,
          }
        }
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          01
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 48,
              },
              Object {
                "color": "#CEE1F2",
              },
              undefined,
            ]
          }
        >
          02
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          03
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          04
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          05
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          06
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          07
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          08
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          09
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          10
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          11
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          12
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          01
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          02
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          03
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          04
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          05
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          06
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          07
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          08
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          09
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          10
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          11
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          12
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          01
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          02
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          03
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          04
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          05
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          06
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          07
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          08
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          09
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          10
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          11
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 75,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#869DB2",
              },
              Object {
                "fontFamily": "Nunito_400Regular",
                "fontSize": 44,
              },
              Object {
                "color": "#869DB2",
              },
              undefined,
            ]
          }
        >
          12
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "height": 62.5,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
